src/main/java/org/javarosa/core/model/condition/Condition.java

```diff
@@ -103,65 +94,26 @@ public boolean isCascadingToChildren() {
         return trueAction.isCascading();
     }
 
-    /**
-     * conditions are equal if they have the same actions, expression, and triggers, but NOT targets or context ref
-     */
-    // TODO Improve this method and simplify
     @Override
     public boolean equals(Object o) {
-        if (o instanceof Condition) {
-            Condition c = (Condition) o;
-            if (this == c)
-                return true;
-
-            boolean result = false;
-            if (c instanceof Triggerable) {
-                Triggerable t = c;
-                if (this == t) {
-                    result = true;
-                } else if (expr.equals(t.getExpr())) {
-
-                    // The original logic did not make any sense --
-                    // the
-                    try {
-                        // resolved triggers should match...
-                        Set<TreeReference> Atriggers = this.getTriggers();
-                        Set<TreeReference> Btriggers = t.getTriggers();
-
-                        result = (Atriggers.size() == Btriggers.size()) &&
-                            Atriggers.containsAll(Btriggers);
-                    } catch (XPathException e) {
-                    }
-                }
-
-            }
-            return (this.trueAction == c.trueAction && this.falseAction == c.falseAction && result);
-        } else {
-            return false;
-        }
+        return o instanceof Condition
+            && super.equals(o)
+            && trueAction == ((Condition) o).trueAction
+            && falseAction == ((Condition) o).falseAction;
     }
 
-    // region External serialization
+    // region External Serialization
 
     @Override
-    @SuppressWarnings("unchecked")
     public void readExternal(DataInputStream in, PrototypeFactory pf) throws IOException, DeserializationException {
-        expr = (IConditionExpr) ExtUtil.read(in, new ExtWrapTagged(), pf);
-        contextRef = (TreeReference) ExtUtil.read(in, TreeReference.class, pf);
-        originalContextRef = (TreeReference) ExtUtil.read(in, TreeReference.class, pf);
-        List<TreeReference> tlist = (List<TreeReference>) ExtUtil.read(in, new ExtWrapList(TreeReference.class), pf);
-        targets = new ArrayList<>(tlist);
+        readExternal(this, in, pf);
```

Comment:
Is there a particular reason to have the superclass have a variant that takes an object as the first param over calling `super.readExternal` and having the same signature?

Reply:
It's just to justify the `readExternal` override in `Recalculate`, but I guess you're right. I want the subtypes to be completely symmetrical respecting to their supertype.

Reply:
I've done much more objectionable things to try to avoid inheritance so I get it. I think in this case keeping it as it was is the right call. 👍